components/lib/dataprotect/src/main/java/mozilla/components/lib/dataprotect/SecureAbove22Preferences.kt

```diff
@@ -0,0 +1,168 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.lib.dataprotect
+
+import android.annotation.TargetApi
+import android.content.Context
+import android.content.Context.MODE_PRIVATE
+import android.content.SharedPreferences
+import android.os.Build
+import android.os.Build.VERSION_CODES.M
+import android.util.Base64
+import androidx.annotation.VisibleForTesting
+import mozilla.components.support.base.log.logger.Logger
+import java.nio.charset.StandardCharsets
+import java.security.GeneralSecurityException
+import java.security.KeyStore
+
+/**
+ * A wrapper around [SharedPreferences] which encrypts contents on supported API versions (23+).
+ * Otherwise, this simply delegates to [SharedPreferences].
+ *
+ * In rare circumstances (such as APK signing key rotation) a master key which protects this storage may be lost,
+ * in which case previously stored values will be lost as well. Applications are encouraged to instrument such events.
+ *
+ * @param context A [Context], used for accessing [SharedPreferences].
+ */
+class SecureAbove22Preferences(context: Context) {
+    companion object {
+        const val KEY_PREFERENCES_POST_M = "key_preferences_post_m"
+        const val KEY_PREFERENCES_PRE_M = "key_preferences_pre_m"
+
+        private const val BASE_64_FLAGS = Base64.URL_SAFE or Base64.NO_PADDING
+    }
+
+    private val keystore: Keystore? = if (Build.VERSION.SDK_INT < M) {
```

Comment:
nit: If you move those `Build.VERSION.SDK_INT` checks behind a method than you could mock that and test the whole migration flow without calling into `migratePrefs()` manually?